doc/developer/design/20220303_secrets.md

```diff
@@ -0,0 +1,264 @@
+        /// The id of the secret to delete.
+        id: GlobalId
+    }
+}
+```
+
+There will be two concrete implementations of the `SecretsController`:
+
+  * The `FilesystemSecretsController` will store secrets on the local filesystem.
+    It will not support multi-operation atomicity, but will support atomically
+    updating or removing one secret at a time.
+
+  * The `KubernetesSecretsController` will store secrets in Kubernetes. It
+    will support multi-operation atomicity.
+
+All secrets controllers expose the contents of secrets to the [`STORAGE`]
+and [`COMPUTE`] layer by way of a directory that contains one file per secret,
+where the name of the file is the secret's ID and the contents of the file are
+the secret's contents. For example, if IDs `u12` and `u36` are secrets, the
+directory would look like this:
+
+```
+secrets/
+  u12
+  u36
+```
+
+This bears repeating: even the `KubernetesSecretsController` will expose the
+contents of the secrets via the filesystem. The filesystem is the *lingua
+franca* for exposing secrets to applications. Kubernetes supports exposing the
+contents of secrets as files on the filesystem, as does [HashiCorp Vault].
+
+The [`STORAGE`] and [`COMPUTE`] layers will need to be taught to re-read the
+contents of secrets that they consume from disk periodically, in case the
+contents have bene updated by an `ALTER SECRET`. We may wish to build a helper
+"secret consumer" library that uses filesystem events to notice when the secrets
+change, rather than polling periodically.
+
+### Transaction integration
+
+Providing proper SQL transaction semantics requires a complicated integration
+between catalog transactions and secret controller transactions.
+
+Let's consider each of the cases in turn:
+
+  * The SQL transaction contains exactly one `CREATE SECRET` statement.
+
+    The coordinator calls `SecretsController::apply` with a single
+    `SecretOp::Ensure`. If the apply fails, the coordinator reports the SQL
+    transaction as failed. If the apply succeeds, the coordinator calls
+    `Catalog::transact` to create the secret object. If the catalog transaction
+    succeeds, the coordinator reports the SQL transaction as successful. If
+    the catalog transaction fails, the coordinator makes a best-effort attempt
+    to remove the secret by calling `SecretsController::apply` with a single
+    `SecretOp::Delete`, then reports the SQL transaction as failed.
+
+  * The SQL transaction contains exactly one `DROP SECRET` statement.
+
+    The coordinator calls `Catalog::transact` to remove the secret from the
+    catalog. If the catalog transaction fails, the coordinator reports the SQL
+    transaction as failed. If the catalog transaction succeeds, the coordinator
+    calls `SecretsController::apply` with a single `SecretOp::Delete` to remove
+    the secret from disk. If the apply succeeds, the coordinator reports the SQL
+    transaction as successful. If the apply fails, the coordinator logs an
```

Comment:
Got it. So we leave the orphaned secret in the controller for later garbage collection.